include/filewriter/filewriter_impl.hpp

```diff
@@ -0,0 +1,91 @@
+/**
+ * @file filewriter_impl.hpp
+ * @author Jeffin Sam
+ */
+#ifndef MLBOARD_FILE_WRITER_IMPL_HPP
+#define MLBOARD_FILE_WRITER_IMPL_HPP
+
+#include "filewriter.hpp"
+
+namespace mlboard {
+
+
+FileWriter::FileWriter(std::string logdir,
+                       int maxQueueSize,
+                       std::size_t flushmilis)
+{
+  const auto p1 = std::chrono::system_clock::now();
+  std::string currentTime =
+      std::to_string(std::chrono::duration_cast<std::chrono::seconds>(
+      p1.time_since_epoch()).count());
+  this->logdir = logdir + "/events.out.tfevents." + currentTime + ".v2";
+  close_ = true;
+  this->flushmilis = flushmilis;
+  size_t &maxSize = this->q.MaxSize();
+  maxSize = maxQueueSize;
+  thread_ = new std::thread(&FileWriter::WriteSummary, this);
+  nexttime =
+      std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+  outfile.open(this->logdir, std::fstream::out |
+      std::ios::trunc | std::ios::binary);
+}
+
+void FileWriter::WriteSummary()
+{
+  // This is a thread that will continously write summary one by one into file.
+  while (true)
+  {
+    // Break the loop if eveything is done.
+    if (!(close_ || q.size() != 0)) break;
+    std::time_t timenow =
+        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+    if (timenow >= nexttime)
+    {
+      while (q.size() > 0)
```

Comment:
Okay, this has just clicked my mind. Basically we are using a mutex to gain control over the queue and to synchronize its access. We are acquiring and releasing the lock multiple times while emptying (in the `ShareQueue::Pop()`). Because of this, two things can happen

1.  it might happen that someone (`the main thread`) pushes into the queue by acquiring lock in the middle of clearing the queue. This might lead to some other issues.
2. It is also not optimum to acquire and release the mutex lock multiple times. 

So, I think it's better that we create a function in `SharedQueue` itself to empty it. A function like `ClearQueue()/ FlushQueue()` would acquire the lock once in the function call and will release it once the queue becomes empty and all objects are written to the file. 

 Let me know what you think about it.

Reply:
Then it will lock the queue for a longer instance and we won't be able to push it to the queue  delaying the training if we are pushing in our callbacks between the training instance

Reply:
Yeah, you are correct. I forgot about the call backs. Let's keep it as it is.